staging/vhost-device-can/src/can.rs

```diff
@@ -0,0 +1,455 @@
+                            flags: frame.id_flags().bits().into(),
+                            sdu: {
+                                let mut sdu_data: [u8; 64] = [0; 64];
+                                sdu_data[..frame.data().len()].copy_from_slice(frame.data());
+                                sdu_data
+                            },
+                        };
+
+                        match controller.push(read_can_frame) {
+                            Ok(_) => warn!("New Can frame was received"),
+                            Err(_) => {
+                                warn!("Error read/push CAN frame");
+                                return Err(Error::SocketRead);
+                            }
+                        }
+                    }
+                    CanAnyFrame::Fd(frame) => {
+                        // CAN FD frame
+                        trace!("Received CAN FD message: {:?}", frame);
+
+                        let read_can_frame = VirtioCanFrame {
+                            msg_type: VIRTIO_CAN_RX.into(),
+                            can_id: frame.raw_id().into(),
+                            length: (frame.data().len() as u16).into(),
+                            reserved: 0.into(),
+                            flags: frame.id_flags().bits().into(),
+                            sdu: {
+                                let mut sdu_data: [u8; 64] = [0; 64];
+                                sdu_data[..frame.data().len()].copy_from_slice(frame.data());
+                                sdu_data
+                            },
+                        };
+
+                        match controller.push(read_can_frame) {
+                            Ok(_) => warn!("New Can frame was received"),
+                            Err(_) => {
+                                warn!("Error read/push CAN frame");
+                                return Err(Error::SocketRead);
+                            }
+                        }
+                    }
+                    CanAnyFrame::Remote(frame) => {
+                        // Remote CAN frame
+                        trace!("Received Remote CAN message: {:?}", frame);
+                        continue;
+                    }
+                    CanAnyFrame::Error(frame) => {
+                        // Error frame
+                        trace!("Received Error frame: {:?}", frame);
+                        continue;
+                    }
+                }
+
+                controller
+                    .rx_event_fd
+                    .write(1)
+                    .expect("Fail to write on rx_event_fd");
+            }
+        }
+    }
+
+    pub(crate) fn exit_read_thread(&mut self) {
+        trace!("Exit can read thread\n");
+        self.status = false;
+    }
+
+    pub(crate) fn config(&mut self) -> &VirtioCanConfig {
+        trace!("Get config\n");
+        if self.busoff {
+            self.config.status = VIRTIO_CAN_S_CTRL_BUSOFF.into();
```

Comment:
The spec only defines the VIRTIO_CAN_S_CTRL_BUSOFF bit, however, I do not think that reflect the status of the device but only:
```
The bit VIRTIO_CAN_S_CTRL_BUSOFF in status is used to indicate the unsolicited CAN controller state
change from started to stopped due to a detected bus off condition.
```
I was wondering if we do not require another bit to indicate the status.

Reply:
The spec states (section 5.20.5.1):
```
If a status update is necessary, the device updates the configuration status before marking the
request used. As the configuration status change is caused by a request from the driver the device is
allowed to omit the configuration change notification here.
```
But I do not understand what it means with `status update`. I do not think it means changing the `busoff` bit.

Reply:
Hello Matias, since that `busoff` story is not crystal clear to me also through
the virtio-spec, I followed the implementation of virtio-can device provided by
OpenSynergy [here](https://github.com/OpenSynergy/qemu/blob/115540168f92ba5351a20b9c62552782ea1e3e04/hw/net/can/can_virtio.c). I believe that is safe to do since 2 out of 3 contributor to
virtio-can-spec (Harald Mommer, Mikhail Golubev-Ciuchea) are also the main
contributors on the virtio-can device on QEMU.

Having that in mind, I will try first to provide my interpretation of the
virtio-spec and then answer in-line to your comments.

Virtio-spec useful points:
a)
```
The bit VIRTIO_CAN_S_CTRL_BUSOFF in status is used to indicate the unsolicited
CAN controller state change from started to stopped due to a detected bus off
condition.
```

b)
```
There are certain error conditions so that the physical CAN controller
has to stop participating in CAN communication on the bus. If such an
error condition occurs the device informs the driver about the
unsolicited CAN controller state change by setting the
VIRTIO_CAN_S_CTRL_BUSOFF bit in the configuration \field{status} field.

After bus-off detection the CAN controller is in STOPPED state. The CAN
controller does not participate in bus communication any more so all CAN
messages pending for transmission are put into the used queue with
\field{result} VIRTIO_CAN_RESULT_NOT_OK.
```

Based on points a), b) and also to the implementation of virtio-can-device in
QEMU provided by Opensynergy ([here](https://github.com/OpenSynergy/qemu/blob/115540168f92ba5351a20b9c62552782ea1e3e04/hw/net/can/can_virtio.c)), I assume that:
1) The referenced `CAN controller` is part of the vhost-user-can device and
   responsible of communicating with the CAN bus located on the host.
2) The `VIRTIO_CAN_S_CTRL_BUSOFF` is set by the vhost-user-can device and
   propagated to the virtio-can driver when the nested `CAN controller`
   changes its status from STARTED to STOPPED unsolicitedly.
    - This `CAN controller` status change can occur in cases like [this](https://github.com/OpenSynergy/qemu/blob/115540168f92ba5351a20b9c62552782ea1e3e04/hw/net/can/can_virtio.c#L244),
      when the CAN message received by the underlying HW, fulfils the condition
      `(can_id & CAN_ERR_FLAG) & (can_id & CAN_ERR_BUSOFF)`.

In-line comments:
> I was wondering if we do not require another bit to indicate the status.

That indeed. I kept the variable `busoff` to be aligned with Opensynergy's
implementation of virtio-can device [here](https://github.com/OpenSynergy/qemu/blob/115540168f92ba5351a20b9c62552782ea1e3e04/hw/net/can/can_virtio.c#L397) but as you correctly noticed,
practically we can directly modify the `config.status` variable. The new
implementation includes that modification.

> Is it OK to change the `busoff` here? The spec states
>
> ```
> The bit VIRTIO_CAN_S_CTRL_BUSOFF in status is used to indicate the unsolicited CAN controller state
> change from started to stopped due to a detected bus off condition.
> ```
>
> So I think the switching to `False` shall happen independent of the driver although I do not know when or how.

> The spec states (section 5.20.5.1):
>
> ```
> If a status update is necessary, the device updates the configuration status before marking the
> request used. As the configuration status change is caused by a request from the driver the device is
> allowed to omit the configuration change notification here.
> ```
>
> But I do not understand what it means with `status update` because the only defined bit in the status is the `busoff` but I do not think it means changing it.

Based on the assumptions 1), 2), I think that since the vhost-user-can device
has the "authority" to change the `config.status` to `VIRTIO_CAN_S_CTRL_BUSOFF`,
it could also reset it to the default value when control messages are received.
That gives the opportunity to the user retry sending CAN messages after
performing `ip link set down can0 && ip link set up can0` on the VM side and
check if the previous `BUSOFF` error is still being received or was a one time error.
The same approach is followed by the virtio-can device in QEMU ([here](https://github.com/OpenSynergy/qemu/blob/115540168f92ba5351a20b9c62552782ea1e3e04/hw/net/can/can_virtio.c#L397) and [here](https://github.com/OpenSynergy/qemu/blob/115540168f92ba5351a20b9c62552782ea1e3e04/hw/net/can/can_virtio.c#L404)).

Please let me know what you think of the above, and if you have something else
in mind on how to treat that `BUSOFF` case.

Last point on my side, is that I have not yet found an obvious way, to trigger
a notification from the device side, when the configuration has been altered.
For the time being, I will keep a note in the code and we can fill that later,
since, practically, that does not have any functional impact.

Reply:
Thanks for explanation! It seems that the `BUSOFF` bit in the status is set to indicate the unsolicited CAN controller state change from started to stopped due to a detected bus off condition. However, I do not know what shall happen if the driver sends a `VIRTIO_CAN_SET_CTRL_MODE_START` and the `BUSOFF` bit is set. The current implementation and the QEMU's one just clear the `BUSOFF` bit when getting `VIRTIO_CAN_SET_CTRL_MODE_START`. Since the spec is not specific about this, I would leave it as it is.

Reply:
>  Since the spec is not specific about this, I would leave it as it is.
Ok great then, thanks, let's leave it as it is for the moment. If in the future the spec become more clear about it, we can always revisite it.

> However, I do not know what shall happen if the driver sends a `VIRTIO_CAN_SET_CTRL_MODE_START` and the `BUSOFF` bit is set.
I think that in this case, we can clear the `BUSOFF` bit (as QEMU does), since  if the actual error persists, the vhost-device-can will try to write to the socket and fail which is going to set the `BUSOFF` bit back to `1`.